transcom/mymove
# B 18092 Update Prime order document download api. New parameter for document type.

## [B-18092](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A871066)
### [E-04799](https://www13.v1host.com/USTRANSCOM38/Epic.mvc/Summary?oidToken=Epic%3A864544)

## Summary

This PR is to finalize the new Prime order document  download API. The REST URI has been updated to accommodate the ability to selectively generate a PDF file by document type (per internal discussions). In addition, the URI has been also changed per best REST API practice. 

`GET: /moves/:locator/documents?type:type` where `:type` can be [ALL, ORDERS, AMENDMENTS]

API is designed to handle the following use cases:

1. merge all documents(order and amendments) into one PDF (type=ALL)
2. merge only order documents into one PDF  (type=ORDERS)
3. merge only amendment documents into one PDF (type=AMENDMENTS)

By default: `GET: /moves/:locator/documents` will be executed as type=ALL

Sample output (locator= AMDORD)
type=ALL
[Customer-ORDERS,AMENDMENTS-for-MTO-AMDORD-2024-01-22T20-51-28.327Z.pdf](https://github.com/transcom/mymove/files/14015997/Customer-ORDERS.AMENDMENTS-for-MTO-AMDORD-2024-01-22T20-51-28.327Z.pdf)

type=ORDERS
[Customer-ORDERS-for-MTO-AMDORD-2024-01-22T22-20-35.169Z.pdf](https://github.com/transcom/mymove/files/14016082/Customer-ORDERS-for-MTO-AMDORD-2024-01-22T22-20-35.169Z.pdf)

type=AMENDMENTS
[Customer-AMENDMENTS-for-MTO-AMDORD-2024-01-22T22-21-14.353Z.pdf](https://github.com/transcom/mymove/files/14016089/Customer-AMENDMENTS-for-MTO-AMDORD-2024-01-22T22-21-14.353Z.pdf)

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

Best way to test is to invoke the API via Postman

1. Find locator that is permissible for download and invoke via Postman (see SQL below) or log into Prime Simulator UI and grab a locator code from one of the list moves.
2. Response will display in Postman view as PDF
3. Save response and open in PDF viewer. Check bookmarks for outline of file labeling what is what -- orders and amendments. File should merge everything together --  PDF and images types.

**GET:  GET: /moves/:locator/documents?type:type**
 where `:type` can be [ALL, ORDERS, AMENDMENTS]

Business Rule Validations:
- Move owner/client's origin duty location does not provide services counseling 

Find locator code where it has n number of order uploaded documents (change count, by default its set as 1) and meets the business rule validations.
```
SELECT mainSubquery.move_id
	,moves.locator
	,moves.STATUS
	,mainSubquery.orders_id
	,mainSubquery.uploaded_orders_id
	,mainSubquery.uploaded_amended_orders_id
FROM PUBLIC.moves moves
JOIN (
	SELECT move_id
		,orders_id
		,uploaded_orders_id
		,uploaded_amended_orders_id
	FROM (
		SELECT downloadable_moves.id move_id
			,downloadable_moves.STATUS
			,downloadable_moves.orders_id
		FROM (
			SELECT moves.*
			FROM PUBLIC.moves moves
			JOIN (
				SELECT orders.id AS order_id
				FROM PUBLIC.orders orders
				JOIN (
					SELECT document_id
					FROM (
						SELECT document_id
							,count(document_id)
						FROM PUBLIC.user_uploads
						GROUP BY document_id
						) t
					WHERE count >= 1
					) doc_cnt_view ON orders.uploaded_orders_id = doc_cnt_view.document_id
				) order_uploaded_view ON moves.orders_id = order_uploaded_view.order_id
			) moves_by_uploaded_doc_cnt
		JOIN (
			SELECT m.*
			FROM (
				SELECT o.*
				FROM PUBLIC.orders o
				JOIN PUBLIC.duty_locations dl ON o.origin_duty_location_id = dl.id
					AND dl.provides_services_counseling = false
				) t
			JOIN PUBLIC.moves m ON t.id = m.orders_id
			) downloadable_moves ON moves_by_uploaded_doc_cnt.id = downloadable_moves.id
		) innerSubquery
	JOIN PUBLIC.orders orders ON orders.id = innerSubquery.orders_id
	) mainSubquery ON moves.id = mainSubquery.move_id;
```
### Backend
